zipkin-storage/cassandra-v1/src/main/java/zipkin2/storage/cassandra/v1/IndexTraceId.java

```diff
@@ -0,0 +1,247 @@
+     * same trace.
+     */
+    @Override public Iterator<Input> iterator() {
+      Set<IndexTraceId.Input> result = entriesThatIncreaseGap();
+      if (inputs.size() > result.size() && LOG.isDebugEnabled()) {
+        int delta = inputs.size() - result.size();
+        LOG.debug("optimized out {}/{} inserts into {}", delta, inputs.size(), table);
+      }
+      return result.iterator();
+    }
+
+    Set<IndexTraceId.Input> entriesThatIncreaseGap() {
+      if (inputs.size() <= 1) return inputs;
+      Set<Entry<String, Long>> toUpdate = new LinkedHashSet<>();
+      Map<Entry<String, Long>, Set<Long>> mappedInputs = new LinkedHashMap<>();
+
+      // Enter a loop that affects shared state when an update widens the time interval for a key.
+      for (IndexTraceId.Input input : inputs) {
+        Entry<String, Long> key =
+          new SimpleImmutableEntry<>(input.partitionKey(), input.trace_id());
+        long timestamp = input.ts();
+        add(mappedInputs, key, timestamp);
+        for (; ; ) {
+          Pair oldRange = partitionKeyAndTraceIdToRange.get(key);
+          if (oldRange == null) {
+            // Initial state is where this key has a single timestamp.
+            oldRange =
+              partitionKeyAndTraceIdToRange.putIfAbsent(key, new Pair(timestamp, timestamp));
+
+            // If there was no previous value, we need to update the index
+            if (oldRange == null) {
+              toUpdate.add(key);
+              break;
+            }
+          }
+
+          long first = Math.min(timestamp, oldRange.left);
+          long last = Math.max(timestamp, oldRange.right);
+
+          Pair newRange = new Pair(first, last);
+          if (oldRange.equals(newRange)) {
+            break; // the current timestamp is contained
+          } else if (partitionKeyAndTraceIdToRange.replace(key, oldRange, newRange)) {
+            toUpdate.add(key); // The range was extended
+            break;
+          }
+        }
+      }
+
+      // When the loop completes, we'll know one of our updates widened the interval of a trace, if
+      // it is the first or last timestamp. By ignoring those between an existing interval, we can
+      // end up with less Cassandra writes.
+      Set<IndexTraceId.Input> result = new LinkedHashSet<>();
+      for (Entry<String, Long> needsUpdate : toUpdate) {
+        Pair range = partitionKeyAndTraceIdToRange.get(needsUpdate);
+        if (containsEntry(mappedInputs, needsUpdate, range.left)) {
+          result.add(Input.create(needsUpdate.getKey(), range.left, needsUpdate.getValue()));
+        }
+        if (containsEntry(mappedInputs, needsUpdate, range.right)) {
+          result.add(Input.create(needsUpdate.getKey(), range.right, needsUpdate.getValue()));
+        }
+      }
+      return result;
+    }
+
+    static <K, V> void add(Map<K, Set<V>> multimap, K key, V value) {
+      Set<V> valueContainer = multimap.get(key);
```

Comment:
Since this is storage, guess we can use Java 8 `computeIfAbsent(key, unused -> new LinkedHashSet<>())`